Add service response data for listing calendar events

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add the capability of response data for for the entity component.

The entity call function supports returning values for multiple entities, however the entity component only supports return values for one entity. (This is somewhat arbitrary, but reduces scope).

Add a service description for the calendar list event service call. Note that a service description of the response is pending.

Work items:
- [X] End to end working
- [ ] More test coverage for events service (e.g. no matches)
- [ ] More test coverage for entity component (no services, too many services, no return values, etc) 
- [ ] End user documentation for return values


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- https://github.com/home-assistant/home-assistant.io/pull/27863
- https://github.com/home-assistant/home-assistant.io/pull/27934

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

allenporter committed:
Add service response data for listing calendar events

Add the capability of response data for for the entity component.

allenporter committed:
Rename input arguments and add service description

balloob commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -743,3 +753,22 @@ async def async_create_event(entity: CalendarEntity, call: ServiceCall) -> None:
+    calendar: CalendarEntity, service_call: ServiceCall
+) -> dict[str, Any]:
+    """Handle snapshot services calls."""
+    if not service_call.return_values:
```
I wonder if this is something we should allow specifying during service registration. It will also help us in the frontend to differentiate between return-only services

allenporter replied:
That's a good idea. If we are going to require that you only ask for return values for services with return values then it makes sense to me.

balloob commented on homeassistant/components/calendar/services.yaml:
```diff
@@ -52,3 +52,22 @@ create_event:
+    entity:
+      domain: calendar
+  fields:
+    start_date_time:
```
I bet that the main use case for this function is going to be to query for next X hours. Wouldn’t it make more sense to take in a duration as a field ?

MartinHjelmare replied:
Looks good, I think.

MartinHjelmare commented on homeassistant/helpers/service.py:
```diff
@@ -672,13 +680,14 @@ def async_set_service_schema(
     """Handle an entity service call.
 
-    Calls all platforms simultaneously.
+    Calls all platforms simultaneously. The service call may request response data which will result in returning
```
Docstrings should not be longer than 72 characters per line.

allenporter committed:
Improve list events to be more user friendly

Allow the end date to be determined based on a relative time duration. Make the start time optional and set to "now". Add additional test coverage. Update demo calendar to actually perform date range checks.

allenporter committed:
Wrap docstrings properly.

allenporter committed:
Increase test coverage

allenporter committed:
Update to use new API calls

allenporter committed:
Readability improvements

allenporter committed:
Wrap docstrings

allenporter commented on homeassistant/helpers/entity_component.py:
```diff
@@ -216,18 +223,33 @@ def async_register_entity_service(
+            if not call.return_response:
+                return None
+            if not response_data:
+                raise HomeAssistantError(
```
I didn't see an easy way to hit this.

MartinHjelmare commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -250,6 +257,20 @@ def _validate_rrule(value: Any) -> str:
 
+SERVICE_LIST_EVENTS: Final = "list_events"
+SERVICE_LIST_EVENTS_SCHEMA: Final = vol.All(
+    cv.has_at_most_one_key(EVENT_END_DATETIME, EVENT_DURATION),
```
If at least one of them is required, use `cv.has_at_least_one_key`.

MartinHjelmare commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -250,6 +257,20 @@ def _validate_rrule(value: Any) -> str:
+        {
+            vol.Optional(EVENT_START_DATETIME): datetime.datetime,
+            vol.Optional(EVENT_END_DATETIME): datetime.datetime,
+            vol.Optional(EVENT_DURATION): vol.All(
```
Use vol.Exclusive here and group duration and eg start time.

MartinHjelmare replied:
Probably use two schemas and combine them with `vol.All` to gather all the requirements.

allenporter replied:
Can you elaborate?  My understanding of this is:
- start is optional
- either duration or end are required, but not both (has at most one key, at least one key)

I think understand the /what/, but not the /why/ and the higher level goal of the suggestion. Are we missing some test cases for failures? or is this simpler code better than the lower level constructs? Is this suggesting a replacement or complementary to what is already here? Generally, while voluptuous has a lot of examples in documentation, I think its missing a user guide especially for more advanced things.

MartinHjelmare replied:
These are the rules I suggest:
If duration is set, start and end should not be set. If start and end is set, duration should not be set. If any of start or end is set, both need to be set. Either duration, or start and end must be set.

To accomplish this we can use `cv.has_at_least_one_key` for duration and one of start or end, `vol.Exclusive` for duration and one of start or end, and `vol.Inclusive` for start and end.

MartinHjelmare replied:
Let's drop the exclusive rule. It's enough to just make sure not both duration and end is set.

MartinHjelmare commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -250,6 +257,20 @@ def _validate_rrule(value: Any) -> str:
+    cv.has_at_most_one_key(EVENT_END_DATETIME, EVENT_DURATION),
+    cv.make_entity_service_schema(
+        {
+            vol.Optional(EVENT_START_DATETIME): datetime.datetime,
```
Start and end time should be inclusive.

allenporter replied:
Start is now made optional, defaulting to now. Should it still be inclusive?  I understand /what/ your comment is saying, but not sure /why/ so want to make sure I understand the point you're making.

MartinHjelmare replied:
Inclusive means optional. If either of start or end is set, both should be set. I think we should only default start to now when using duration.

allenporter replied:
I see, i made it optional on purpose since it seemed unnecessarily rigid to make that distinction. I get why it makes sense conceptually to have start and end paired, but in practice it seems nice to not have to make users care about it if they want to just use a duration or switch to an end time later.

Do you have much conviction about it? I can adjust if you do, but just wanted to check one more time.

MartinHjelmare replied:
Ok, we can drop the inclusive.

MartinHjelmare commented on homeassistant/helpers/service.py:
```diff
@@ -804,8 +812,10 @@ async def entity_service_call(  # noqa: C901
-        future.result()  # pop exception if have
+
+    response_data: dict[str, ServiceResponse] = {}
+    for entity, task in zip(entities, done):
```
Why should we support multiple entities here for the response when we don't do it in general. Can't we raise already here for incorrect call?

allenporter replied:
Fixed.

allenporter committed:
Require at least one of end or duration

allenporter committed:
Check for multiple entity matches earlier in the request

MartinHjelmare commented on homeassistant/helpers/service.py:
```diff
@@ -672,13 +678,15 @@ def async_set_service_schema(
 
-    Calls all platforms simultaneously.
+    Calls all platforms simultaneously. The service call may request
+    response data which will return a dictionary of entity
```
We need to update the docstring.

MartinHjelmare commented on homeassistant/components/calendar/__init__.py:
```diff
@@ -743,3 +770,21 @@ async def async_create_event(entity: CalendarEntity, call: ServiceCall) -> None:
+async def async_list_events_service(
+    calendar: CalendarEntity, service_call: ServiceCall
+) -> ServiceResponse:
+    """Handle snapshot services calls."""
```
Stale docstring.

allenporter committed:
Update documentation strings
